File: fap_online_backend/fap_online_backend/src/main/java/com/example/demo/service/AdminRoleService.java
package com.example.demo.service;

import com.example.demo.dto.AdminRoleDTO;
import com.example.demo.dto.AdminRoleRequest;
import com.example.demo.dto.AdminUserDTO;
import com.example.demo.dto.AssignRoleRequest;

import java.util.List;

public interface AdminRoleService {

    List<AdminRoleDTO> getRoles();

    AdminRoleDTO getRoleById(Integer roleId);

    AdminRoleDTO createRole(AdminRoleRequest request);

    AdminRoleDTO updateRole(Integer roleId, AdminRoleRequest request);

    void deleteRole(Integer roleId);

    AdminUserDTO assignRole(AssignRoleRequest request);

    AdminUserDTO unassignRole(AssignRoleRequest request);
}

File: fap_online_ui/lib/features/admin/data/services/admin_role_service.dart
import '../../../../services/api_service.dart';
import '../../../../config/api_endpoints.dart';
import '../models/admin_models.dart';

class AdminRoleService {
  final ApiService _api = ApiService();

  Future<List<AdminRoleModel>> getRoles() async {
    final data = await _api.get(ApiEndpoints.adminRoles);
    return (data as List<dynamic>)
        .map((e) => AdminRoleModel.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<AdminRoleModel> createRole(AdminRoleRequest request) async {
    final data = await _api.post(ApiEndpoints.adminRoles, request.toJson());
    return AdminRoleModel.fromJson(data as Map<String, dynamic>);
  }

  Future<AdminRoleModel> updateRole(int roleId, AdminRoleRequest request) async {
    final data = await _api.put(
      ApiEndpoints.adminRoleById(roleId),
      body: request.toJson(),
    );
    return AdminRoleModel.fromJson(data as Map<String, dynamic>);
  }

  Future<void> deleteRole(int roleId) async {
    await _api.delete(ApiEndpoints.adminRoleById(roleId));
  }

  Future<AdminUserModel> assignRole(AssignRoleRequest request) async {
    final data =
        await _api.post(ApiEndpoints.adminRoleAssign, request.toJson());
    return AdminUserModel.fromJson(data as Map<String, dynamic>);
  }

  Future<AdminUserModel> unassignRole(AssignRoleRequest request) async {
    final data =
        await _api.post(ApiEndpoints.adminRoleUnassign, request.toJson());
    return AdminUserModel.fromJson(data as Map<String, dynamic>);
  }
}
